### PR TITLE
gprofiler/python_ebpf.py: non-blocking standard output streams reading

### PR DESCRIPTION
**Pull request by roi-granulate**
## Checklist:

- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have updated the relevant documentation.
- [ ] I have added tests for new logic.

### CODE REVIEW

File: gprofiler/profilers/python_ebpf.py [hunk 1]
```diff
@@ -78,6 +79,7 @@ def __init__(
     ):
         super().__init__(frequency, duration, profiler_state)
         self.process: Optional[Popen] = None
+        self.process_selector: Optional[selectors.DefaultSelector] = None
```
Comment (d3dave) [inline gprofiler/profilers/python_ebpf.py:82]:
```suggestion
        self.process_selector: Optional[selectors.BaseSelector] = None
```

File: gprofiler/profilers/python_ebpf.py [hunk 2]
```diff
@@ -213,6 +215,36 @@ def start(self) -> None:
...
+    def _register_process_selectors(self) -> None:
+        self.process_selector = selectors.DefaultSelector()
+        assert self.process_selector and self.process and self.process.stdout and self.process.stderr  # for mypy
+        self.process_selector.register(self.process.stdout, selectors.EVENT_READ)
+        self.process_selector.register(self.process.stderr, selectors.EVENT_READ)
+
+    def _unregister_process_selectors(self) -> None:
+        assert self.process_selector
+        for selector_key in list(self.process_selector.get_map().values()):
+            self.process_selector.unregister(selector_key.fileobj)
```
Comment (d3dave) [inline gprofiler/profilers/python_ebpf.py:229]: Is there a reason to keep the `self.process_selector` alive and not set it to `None`?

File: gprofiler/profilers/python_ebpf.py [hunk 3]
```diff
@@ -213,6 +215,36 @@ def start(self) -> None:
...
+        """
+        stdout: Optional[str] = None
+        stderr: Optional[str] = None
+        assert self.process_selector and self.process
+        for key, _ in self.process_selector.select(timeout=0):
+            output = key.fileobj.read1(self._OUTPUT_READ_SIZE)  # type: ignore
+            output = cast(str, output)
+            if key.fileobj is self.process.stdout:
+                stdout = output
+            if key.fileobj is self.process.stderr:
```
Comment (d3dave) [inline gprofiler/profilers/python_ebpf.py:245]:
```suggestion
            elif key.fileobj is self.process.stderr:
```

**Review by d3dave — CHANGES_REQUESTED** [review]

File: gprofiler/profilers/python_ebpf.py [hunk 4]
```diff
@@ -213,6 +215,37 @@ def start(self) -> None:
...
+    def _register_process_selectors(self) -> None:
+        self.process_selector = selectors.DefaultSelector()
+        assert self.process_selector and self.process and self.process.stdout and self.process.stderr  # for mypy
+        self.process_selector.register(self.process.stdout, selectors.EVENT_READ)
+        self.process_selector.register(self.process.stderr, selectors.EVENT_READ)
+
+    def _unregister_process_selectors(self) -> None:
+        assert self.process_selector
+        for selector_key in list(self.process_selector.get_map().values()):
+            self.process_selector.unregister(selector_key.fileobj)
```
Comment (d3dave) [inline gprofiler/profilers/python_ebpf.py:229]: I think you can just do `self.process_selector.clear()`

**Review by d3dave — CHANGES_REQUESTED** [review]

File: gprofiler/profilers/python_ebpf.py (hunk 4 above)
Comment (roi-granulate) [reply gprofiler/profilers/python_ebpf.py:229]: you mean `.close()`

**Review by d3dave — APPROVED** [review]

**Pull request merged**